ssl/handshake_server.cc

```diff
@@ -291,13 +293,20 @@ bool ssl_parse_client_cipher_list(
       return false;
     }
 
+    // Storing only supported ciphers
     const SSL_CIPHER *c = SSL_get_cipher_by_value(cipher_suite);
     if (c != NULL && !sk_SSL_CIPHER_push(sk.get(), c)) {
       OPENSSL_PUT_ERROR(SSL, SSL_R_ERROR_IN_RECEIVED_CIPHER_LIST);
       return false;
     }
   }
 
+  // Store raw bytes for cipher suites offered
+  char * tmp = (char *) OPENSSL_malloc(all_ciphers_len + 1);
```

Comment:
* Is extending it with a null byte and the re-computation of `all_ciphers_len` necessary?
* Could this just use `OPENSSL_memdup`:
```c++
  char* tmp =  static_cast<char *>(OPENSSL_memdup(
      client_hello->cipher_suites, client_hello->cipher_suites_len));
```
* How should we handle `tmp` being NULL after malloc?

Reply:
As long as we treat it as a string and return strlen in SSL_client_hello_get0_ciphers, it's probably best to explicitly set the null char and compute length. CBS_data(&cipher_suites) returns the relevant data as a char pointer but doesn't zero out the memory after it/has a null char at the end. Using memdup works to copy the data but there is no guarantee of there being a null char at the end. We would need to allocate another byte and then explicitly set it. 

To handle tmp being NULL, I'll add a failure since this allocation shouldn't fail in the handshake.